internal/gw/gw.go

```diff
@@ -63,6 +61,16 @@ type Client interface {
 	//
 	// This function is intended for debugging purposes only.
 	WhoAmI(context.Context) (map[string]interface{}, error)
+
+	// GetUploadItems determines a list of upload items by querying existing items in the
+	// S3 bucket.
+	//
+	// For each item, onPresent is invoked if the item was already present prior to the call,
+	// onDuplicate is invoked if a duplicate item is detected in the publish.
+	GetUploadItems(ctx context.Context, items []walk.SyncItem,
```

Comment:
With this design, we can no longer uploaded *any* item until after we've finished checking the presence of *all* items.  If you imagine for instance we have 500,000 items, for the very first item we can very quickly determine that it needs to be uploaded (and therefore put it on the return value we're building up), but we still need to do up to 499,999 more HEAD queries to S3 before we get started on uploading it.  The unnecessary delay is not ideal and I suppose it could end up wasting a large amount of time for some publishes.

To avoid this problem, the interface probably shouldn't have this method added.  I think it'd make more sense if EnsureUploaded was updated to have all three of the (uploaded/present/duplicate) callbacks and still just return an error, so we do not have any externally visible "calculate what needs to be uploaded, then upload" steps.  Then the implementation internally can start uploading each item as soon as it's known that it should be uploaded.